lsコマンドを作る1

pofkuma opened the pull request:
オプション無しのlsコマンドを作成しました。

ご確認よろしくお願いします。

pofkuma committed:
Add initial test

テストコードを作成

pofkuma committed:
Add initial program

とりあえず動くコードができた

pofkuma committed:
Modify test

テストコードを修正

pofkuma committed:
Fix bug and Delete commment

バグ修正と不要なコメントの削除

pofkuma committed:
Improve some codes

実装の改善と、テストの修正

pofkuma committed:
Extract subproblems into methods

個別のタスクをメソッドに分割

pofkuma committed:
Make variable names more exact

変数名をより正確な名前にする

summertree128 commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+CURRENT_DIRECTORY = '.'
```
これは定数化しなくても良いかなと個人的には思いました。（もちろん、してダメというわけではないです）

理由としては、
- カレントディレクトリが`.`であることは業界における共通知識のようなものだし、OSなどの環境にも依らないため
- カレントディレクトリが`.`であることは少なくともしばらくの間変わることはなさそうなため
- `CURRENT_DIRECTORY`よりも`'.'`の方がだいぶ短いため

こういう場合、定数化するメリットが薄いので僕だったらしないと思いました。ご参考までに〜

pofkuma replied:
詳しい理由の説明までありがとうございます🙏
定数化する必要性がないと思いましたので、`.`をそのまま使うようにしました。
https://github.com/pofkuma/ruby-practices/pull/6/commits/0e4af3e6ccd351fe255b9f8aa4fdf0cda8433d09

summertree128 commented on 05.ls/ls_test.rb:
```diff
@@ -0,0 +1,55 @@
+require_relative 'ls'
+
+class LsTest < Minitest::Test
+  CONTENT_NAMES_WHICH_BEGIN_WITH_A_DOT = %w[. ..].freeze
```
これも定数化しなくても良いかなと思いました。理由は `CURRENT_DIRECTORY` とほぼ同じです。
（もちろん、しても良いとは思います）

pofkuma replied:
`CURRENT_DIRECTORY`と同様（しかもかなり長い）のでリテラルのままにしました🙏
https://github.com/pofkuma/ruby-practices/pull/6/commits/0e4af3e6ccd351fe255b9f8aa4fdf0cda8433d09

summertree128 commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,46 @@
+      lines[index] += content.ljust(max_name_length + PADDING_SIZE)
+    end
+  end
+  lines.map(&:rstrip).join("\n") << "\n"
```
個人的にはですが、最後の改行は呼び出し側に任せれば良いかなと思いました。（つまり、`print`の代わりに`puts`すればわざわざここで最後に改行文字を追加しなくとも良い）

pofkuma replied:
すみません、こちらコメントすべきだったのですが、
`puts`で改行すると、MInitestで`list_directory_contents`を呼び出すときの返り値としてassert_equalができなかったため、あえて`print`を使っておりました。
テストのやり方を変えた方が良いのでしょうか？
アドバイスいただけると有難いです🙇‍♀️

summertree128 replied:
@pofkuma 
なるほどですねー。であればこのままでも良いですし、ヒアドキュメントの末尾から改行を取り除きたい場合は`chomp`が使えます。
参考：https://secret-garden.hatenablog.com/entry/2015/07/23/124726

pofkuma replied:
@summertree128 
ヒアドキュメントのラベルで`chomp`が使えるんですね。
教えてくださりありがとうございます！
https://github.com/pofkuma/ruby-practices/pull/6/commits/86ba451a38e136027e0f6c301f2a5a17710b31f9 にて修正しました。

summertree128 commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,46 @@
+
+def list_directory_contents(contents)
+  filterd_contents = contents.filter_by_name
+  return if filterd_contents.size.zero?
+
+  filterd_contents.sort.format
```
早期 return は適切に使えばとても効果的な書き方ですが、今回のように return の後の行数が短いのであれば普通の if (unless) の方がスッキリするかなと思いました。修正必須というわけではないですが、参考までに。

```suggestion
  filterd_contents.sort.format unless filterd_contents.size.zero?
```

pofkuma replied:
提案いただいたコードの方が1行でスッキリして良いと思ったので、適用しました🙏

summertree128 commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,46 @@
+
+CURRENT_DIRECTORY = '.'
+PADDING_SIZE = 7
+MAX_COLUMUNS = 3
```
ここを変えれば3列ではなく5列とか100列にすることもできるということですね。ナイスです 👍

summertree128 commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,46 @@
+public :format, :filter_by_name, :list_directory_contents
+
+if $PROGRAM_NAME == __FILE__
+  entries = Dir.entries(CURRENT_DIRECTORY)
```
このままでも動作に問題はないので良いのですが、用途的には`Dir.glob`の方が適しているというか、色々楽に書けるかなと思いました。一度調べてみると良いかもしれません。

pofkuma replied:
ありがとうございます！
`Dir.glob`の方が便利だと思ったので、そちらを使うようにしました。https://github.com/pofkuma/ruby-practices/pull/6/commits/8d06d1eb5f6489333399f8128808168bf3ec7959

[Dir\.\[\] \(Ruby 3\.0\.0 リファレンスマニュアル\)](https://docs.ruby-lang.org/ja/latest/method/Dir/s/=5b=5d.html)

summertree128 reviewed (commented):
レビューしました。

全体的にとても良いと思いました！
まずテストがあるのが良いです。lsでテストを書く人はあまりいないので本当に素晴らしいことだと思います。褒めても褒め足りないくらいです ✨ 
メソッドもうまく、分かりやすく切られていると思いますし処理の内容も良いと思います。

より良くできそうな点や気になった点をいくつかコメントしていますのでご確認ください。
また、プラクティスの方に記載したコメントの方も確認していただいて、対応する修正をしていただきたいです 🙏

pofkuma committed:
Remove constants not needed

定数化する必要性が低いためリテラルのまま使う

pofkuma committed:
Update 05.ls/ls.rb

Co-authored-by: Natsuki Inoue <natsukinoue530000@gmail.com>

pofkuma committed:
Compact return statement

returnを省略して簡潔にする

pofkuma committed:
Use "grob" method

Dir.globの方が便利なのでそちらを使う
ドットで始まるものは除外されるので、今回のテストでは'.'と'..'は取り除かれた前提とする
('-a'オプションに対応するときに考慮する)

pofkuma committed:
Merge branch 'my-ls1' of https://github.com/pofkuma/ruby-practices into my-ls1

pofkuma committed:
Do not use top-level self as a receiver

トップレベルで定義したメソッドを無理矢理publicにしない
(どのオブジェクトでも呼び出せることになり、意図しない呼び出し方を許してしまう)
オブジェクトを引数で受け取るようにする
https://docs.ruby-lang.org/ja/latest/class/main.html

pofkuma commented:
レビューありがとうございます！

> より良くできそうな点や気になった点をいくつかコメントしていますのでご確認ください。 また、プラクティスの方に記載したコメントの方も確認していただいて、対応する修正をしていただきたいです 🙏

[こちらの質問](https://github.com/pofkuma/ruby-practices/pull/6#discussion_r749900179)以外の部分は、一通り修正を行いました。
お手数ですが、再度ご確認をお願いします🙇‍♀️

pofkuma committed:
Use puts instead of print

改行文字の入力を省けるためputsメソッドを使う

テストコードではヒアドキュメントの末尾の改行を取り除く

summertree128 commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,36 @@
+MAX_COLUMUNS = 3
+
+def calculate_line_count(size)
+  line_count = size / MAX_COLUMUNS
+  line_count += 1 unless (size % MAX_COLUMUNS).zero?
```
最初のレビューでは見落としていましたが、割り算の結果を小数として出して、小数点を切り上げすれば一行で同じ結果が得られそうですね。興味があれば次のプラクティスあたりで試してみてください。

pofkuma replied:
@summertree128 
ありがとうございます、次のプラクティスで試してみます💡

summertree128 reviewed (approved):
修正ありがとうございました。レビューOKです 👌 
一点だけ追加コメントしてありますのでご確認をお願いします〜
